src/expr/src/scalar/func.rs

```diff
@@ -1365,18 +1365,51 @@ fn ascii<'a>(a: Datum<'a>) -> Datum<'a> {
     }
 }
 
+/// A timestamp with only a time component.
+pub trait TimeLike: Copy + chrono::Timelike + for<'a> Into<Datum<'a>> {
+    fn extract_hour(&self) -> f64 {
+        f64::from(self.hour())
+    }
+
+    fn extract_minute(&self) -> f64 {
+        f64::from(self.minute())
+    }
+
+    fn extract_second(&self) -> f64 {
+        let s = f64::from(self.second());
+        let ns = f64::from(self.nanosecond()) / 1e9;
+        s + ns
+    }
+
+    fn extract_millisecond(&self) -> f64 {
+        let s = f64::from(self.second() * 1_000);
+        let ns = f64::from(self.nanosecond()) / 1e6;
+        s + ns
+    }
+
+    fn extract_microsecond(&self) -> f64 {
+        let s = f64::from(self.second() * 1_000_000);
+        let ns = f64::from(self.nanosecond()) / 1e3;
+        s + ns
+    }
+}
+
+impl TimeLike for chrono::NaiveTime {}
```

Comment:
If there's a more concise method of these 3 lines, let me know.

Reply:
This is about as concise as it gets with Rust. :)